client/comms/wsconn.go

```diff
@@ -0,0 +1,354 @@
+	return atomic.AddUint64(&conn.rID, 1)
+}
+
+// close terminates all websocket processes and closes the connection.
+func (conn *WsConn) close() {
+	conn.wsMtx.Lock()
+	defer conn.wsMtx.Unlock()
+
+	if conn.ws == nil {
+		return
+	}
+
+	msg := websocket.FormatCloseMessage(websocket.CloseNormalClosure, "")
+	conn.ws.WriteControl(websocket.CloseMessage, msg,
+		time.Now().Add(writeWait))
+	conn.ws.Close()
+}
+
+// connect attempts to establish a websocket connection.
+func (conn *WsConn) connect() error {
+	url := url.URL{
+		Scheme: "wss",
+		Host:   conn.cfg.Host,
+		Path:   conn.cfg.Path,
+	}
+
+	dialer := &websocket.Dialer{
+		Proxy:            http.ProxyFromEnvironment,
+		HandshakeTimeout: 10 * time.Second,
+		TLSClientConfig:  conn.tlsCfg,
+	}
+
+	ws, _, err := dialer.Dial(url.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	ws.SetPingHandler(func(string) error {
+		conn.wsMtx.Lock()
+		defer conn.wsMtx.Unlock()
+
+		now := time.Now()
+		err := ws.SetReadDeadline(now.Add(conn.cfg.PingWait))
+		if err != nil {
+			log.Errorf("read deadline error: %v", err)
+			return err
+		}
+
+		// Respond with a pong.
+		err = ws.WriteControl(websocket.PongMessage, []byte{}, now.Add(writeWait))
+		if err != nil {
+			log.Errorf("pong error: %v", err)
+			return err
+		}
+
+		return nil
+	})
+
+	conn.wsMtx.Lock()
+	conn.ws = ws
+	conn.wsMtx.Unlock()
+
+	return nil
+}
+
+// read fetches and parses incoming messages for processing. This should be
+// run as a goroutine.
+func (conn *WsConn) read() {
+	defer conn.wg.Done()
+
+	for {
+		msg := new(msgjson.Message)
+
+		conn.wsMtx.Lock()
+		ws := conn.ws
+		conn.wsMtx.Unlock()
+
+		err := ws.ReadJSON(msg)
+		if err != nil {
+			if _, ok := err.(*json.UnmarshalTypeError); ok {
+				// JSON decode errors are not fatal, log and proceed.
+				log.Errorf("json decode error: %v", err)
+				continue
+			}
+
+			if websocket.IsCloseError(err, websocket.CloseGoingAway,
+				websocket.CloseNormalClosure) ||
+				strings.Contains(err.Error(), "websocket: close sent") {
+				return
+			}
+
+			if opErr, ok := err.(*net.OpError); ok {
+				if opErr.Op == "read" {
+					if strings.Contains(opErr.Err.Error(),
+						"use of closed network connection") {
+						return
+					}
+				}
+			}
+
+			// Log all other errors and trigger a reconnection.
+			log.Errorf("read error: %v", err)
+			conn.reconnectCh <- struct{}{}
+			return
+		}
+
+		conn.readCh <- msg
+	}
+}
+
+// keepAlive maintains an active websocket connection by reconnecting when
+// the established connection is broken. This should be run as a goroutine.
+func (conn *WsConn) keepAlive() {
+	for {
+		select {
+		case <-conn.reconnectCh:
+			conn.setConnected(false)
+
+			reconnects := atomic.AddUint64(&conn.reconnects, 1)
+			if reconnects > 1 {
+				conn.close()
+			}
+
+			err := conn.connect()
+			if err != nil {
+				log.Errorf("connection error: %v", err)
+
+				go func() {
+					// Attempt to reconnect.
+					time.Sleep(conn.cfg.PingWait)
```

Comment:
`PingWait` might be a little long here, and the semantics are a little confusing. Maybe okay for now though.